### PR TITLE
Add support for System.Net.Sockets.Socket.Disconnect on Unix

### PR DESCRIPTION
**Pull request by luhenry**
This copies the way it is done in Mono (see https://github.com/mono/mono/blob/65025fe/mono/metadata/w32socket-unix.c#L928)

### CODE REVIEW

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 1]
```diff
@@ -1556,12 +1556,22 @@ public static SocketError AcceptAsync(Socket socket, SafeCloseSocket handle, Saf
...
         }
 
         internal static SocketError Disconnect(Socket socket, SafeCloseSocket handle, bool reuseSocket)
         {
-            throw new PlatformNotSupportedException(SR.net_sockets_disconnect_notsupported);
+            using (Socket newsocket = new Socket(socket.AddressFamily, socket.SocketType, socket.ProtocolType))
+            {
+                Interop.Error err = Interop.Sys.Dup2(newsocket.SafeHandle, socket.SafeHandle);
+                return err == Interop.Error.SUCCESS ? SocketError.Success : GetSocketErrorForErrorCode(err);
+            }
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs]: This ends up having the same problem that Mono's connect functionality has in general: the new socket handle isn't configured with any options that may have been set on the original socket, either via Socket's API or via getting Socket's Handle and doing direct configuration on the file descriptor via one's own P/Invokes.  If we want to enable this, we'll need to incorporate something along the lines of what was done in https://github.com/dotnet/corefx/pull/16373.  It's possible most of that should come "for free", since the {Begin}Connect{Async} methods are now instrumented to call ReplaceHandleIfNecessaryAfterFailedConnect() which handles the creation of a new socket and the propagation (where possible) of settings to it, so it may be that little else is needed here other than closing the old handle; I've not tried, though.

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs [hunk 2]
```diff
@@ -67,16 +65,14 @@ public void Disconnect_Success()
...
-                    completed.WaitOne();
+                    if (client.ConnectAsync(args))
+                        completed.WaitOne();
                     Assert.Equal(SocketError.Success, args.SocketError);
                 }
             }
         }
 
         [Fact]
-        [OuterLoop("Issue #11345")]
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: Why did you remove the Outerloop on it?

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs [hunk 3]
```diff
@@ -97,22 +93,20 @@ public void DisconnectAsync_Success()
                     completed.WaitOne();
                     Assert.Equal(SocketError.Success, args.SocketError);
 
-                    Assert.True(client.DisconnectAsync(args));
-                    completed.WaitOne();
+                    if (client.DisconnectAsync(args))
+                        completed.WaitOne();
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: Nit: can you add braces around these newly indented lines, e.g.
```C#
if (client.DisconnectAsync(args))
{
    completed.WaitOne();
}
```

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs [hunk 4]
```diff
@@ -97,22 +93,20 @@ public void DisconnectAsync_Success()
...
-                    completed.WaitOne();
+                    if (client.ConnectAsync(args))
+                        completed.WaitOne();
                     Assert.Equal(SocketError.Success, args.SocketError);
                 }
             }
         }
 
         [Fact]
-        [OuterLoop("Issue #11345")]
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: Why was OuterLoop removed?

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs [hunk 5]
```diff
@@ -1556,12 +1556,22 @@ public static SocketError AcceptAsync(Socket socket, SafeCloseSocket handle, Saf
 
         internal static SocketError DisconnectAsync(Socket socket, SafeCloseSocket handle, bool reuseSocket, DisconnectOverlappedAsyncResult asyncResult)
         {
-            throw new PlatformNotSupportedException(SR.net_sockets_disconnect_notsupported);
+            SocketError socketError = Disconnect(socket, handle, reuseSocket);
+            if (socketError == SocketError.Success)
+            {
+                asyncResult.CompletionCallback(SocketError.Success);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs]: What if it's not socketError == SocketError.Success?  When will the asyncResult be completed in that case?  Seems like in the case of an error, this would cause a hang, as the asyncresult will never be completed.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Unix.cs [hunk 6]
```diff
@@ -122,12 +122,16 @@ internal unsafe SocketError DoOperationConnect(Socket socket, SafeCloseSocket ha
 
         internal SocketError DoOperationDisconnect(Socket socket, SafeCloseSocket handle)
         {
-            throw new PlatformNotSupportedException(SR.net_sockets_disconnect_notsupported);
+            SocketError socketError = SocketPal.Disconnect(socket, handle, _disconnectReuseSocket);
+            if (socketError != SocketError.IOPending)
+            {
+                FinishOperationSync(socketError, 0, SocketFlags.None);
+            }
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Unix.cs]: And what if it does == IOPending?  Or maybe given the implementation, we expect it can't ever be IOPending? In that case this should be changed to something like:
```C#
Debug.Assert(socketError != SocketError.IOPending, "Disconnect on Unix should never be pending");
FinishOperationSync(socketError, 0, SocketFlags.None);
return socketError;
```

File: src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs [hunk 7]
```diff
@@ -67,16 +65,14 @@ public void Disconnect_Success()
 
                     args.RemoteEndPoint = server2.EndPoint;
 
-                    Assert.True(client.ConnectAsync(args));
-                    completed.WaitOne();
+                    if (client.ConnectAsync(args))
```
Comment (stephentoub) [inline src/System.Net.Sockets/tests/FunctionalTests/DisconnectTest.cs]: Do we have any tests for `client.Disconnect(false)`?  If not, we should add one, and make sure that the Unix implementation respects the reuseSocket == false in a way similar to how Windows would (or, alternatively, throw a PlatformNotSupportedException for that restricted case if we can't support it, but it seems like that should be the easy case).

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Unix.cs [hunk 8]
```diff
@@ -122,12 +122,16 @@ internal unsafe SocketError DoOperationConnect(Socket socket, SafeCloseSocket ha
...
+            return socketError;
         }
 
         private void InnerStartOperationDisconnect()
         {
-            throw new PlatformNotSupportedException(SR.net_sockets_disconnect_notsupported);
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Unix.cs]: If this resource isn't used any more, it should be deleted from the Strings.resx.

**Comment by stephentoub** [comment]
cc: @geoffkizer

File: src/Common/src/Interop/Unix/System.Native/Interop.Dup2.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_Dup2", SetLastError = true)]
+        internal static extern Error Dup2(SafeHandle oldfd, SafeHandle newfd);
+    }
+}
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Native/Interop.Dup2.cs]: Is all of this dup2 stuff that was added still needed, or can it be undone?  Seems like with the revised approach, it's no longer necessary.

File: src/Common/src/System/Net/SafeCloseSocket.Unix.cs [hunk 10]
```diff
@@ -161,6 +161,13 @@ public int SendTimeout
             }
         }
 
+        internal bool IsDisconnected { get; private set; } = false;
+
+        internal void SetToDisconnected()
+        {
+            IsDisconnected = true;
+        }
```
Comment (stephentoub) [inline src/Common/src/System/Net/SafeCloseSocket.Unix.cs]: Socket already has an _isDisconnected field; can we just check that rather than adding another field (the implicit one backing IsDisconnected above) to track the same state?  Or is there something different/special about this one?

File: src/System.Net.Sockets/src/System/Net/Sockets/DisconnectOverlappedAsyncResult.Unix.cs [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace System.Net.Sockets
+{
+    // DisconnectOverlappedAsyncResult - used to take care of storage for async Socket BeginAccept call.
+    internal sealed partial class DisconnectOverlappedAsyncResult : BaseOverlappedAsyncResult
+    {
+        public void CompletionCallback(SocketError errorCode)
+        {
+            base.CompletionCallback(0, errorCode);
+        }
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/DisconnectOverlappedAsyncResult.Unix.cs]: Alternatively, we could just make the base CompletionCallback method `protected internal` instead of `protected` (or even just `internal` or `public`, since the type itself is already `internal`), and avoid needing to add another method.

File: src/Common/src/System/Net/SafeCloseSocket.Unix.cs (hunk 10 above)
Comment (luhenry) [inline src/Common/src/System/Net/SafeCloseSocket.Unix.cs]: This doesn't have the same lifetime and accessibility. Socket._isDisconnected can be set to true or false from SetToConnected or SetToDisconnected, while SafeCloseSocket.IsDisconnected can only be set to true from SetToDisconnected. This is because on Unix, once you have disconnected a socket, you can't reuse it (and it gets replaced anyway)

Comment (stephentoub) [inline src/Common/src/System/Net/SafeCloseSocket.Unix.cs]: Accessibility is an implementation detail; as long as it's not public/protected (which it's not), we can change it to meet our needs.

Lifetime is an issue, but I'm not understanding yet why the lifetime is different.  The purpose of these changes is to make the Unix behavior as close to the Windows behavior as possible.  Why do we need this tracked on SafeCloseSocket at all?  Can't we just use the tracking on Socket itself?  If reuseSocket is false, then the handle won't be replaced, and a subsequent attempt to connect will fail because the old handle that already had a connection attempt will still be there.

What am I missing?

**Comment by stephentoub** [comment]
Thanks, @luhenry.  This looks good and clean to me.  I know we had a few iterations on it... are you happy with where it landed?

**Comment by luhenry** [comment]
Thank you @stephentoub , absolutely no problem for the iterations, we all want corefx to be better and available more places! :)

**Comment by stephentoub** [comment]
@geoffkizer, could you take a peek as well before we merge this?

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs [hunk 12]
```diff
@@ -23,7 +23,7 @@ public BaseOverlappedAsyncResult(Socket socket, Object asyncState, AsyncCallback
             if (NetEventSource.IsEnabled) NetEventSource.Info(this, socket);
         }
 
-        protected void CompletionCallback(int numBytes, SocketError errorCode)
+        internal void CompletionCallback(int numBytes, SocketError errorCode)
```
Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs]: Why did this change to internal?

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs]: My feedback: https://github.com/dotnet/corefx/pull/17000#discussion_r106207808

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs]: The issue is, generally it would be very bad to accidentally call CompletionCallback on the base class instead of the derived one.  In this case it doesn't matter because we just defer to the base class.  But I do think it's worth maintaining the same pattern here as everywhere else.

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs]: Meaning... you think we should revert to what @luhenry had initially, with this protected, making the derived type partial, adding to the Unix-specific version of it a method that just delegates to the base, etc.?

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs]: Yes.

At some point we should try to get rid of the need for these partial AsyncResult classes, but until then we should just keep the existing pattern.

**Comment by geoffkizer** [comment]
Aside from the one AsyncResult comment above, this looks good to me.  Thanks for doing this!

File: src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs (hunk 12 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs]: Ok.  I'm going to merge the change and then I'll fix this as part of the SendPacketsAsync change.

Comment (geoffkizer) [inline src/System.Net.Sockets/src/System/Net/Sockets/BaseOverlappedAsyncResult.Unix.cs]: Sounds good to me.  Sorry to have not caught this earlier when you initially made the comment.

**Comment by luhenry** [comment]
@geoffkizer @stephentoub thanks for getting this merged! 👍
